lazysnap/src/main/java/com/google/accompanist/lazysnap/SnappingFlingBehavior.kt

```diff
@@ -0,0 +1,460 @@
+            // If we don't have a valid distance, return the current item
+            return currentItem.index
+        }
+
+        val maximumFlingDistance = maximumFlingDistance(lazyListState.layoutInfo)
+        val flingDistance = decayAnimationSpec.calculateTargetValue(
+            initialValue = -currentItem.offset.toFloat(),
+            initialVelocity = initialVelocity,
+        ).coerceIn(-maximumFlingDistance.toFloat(), maximumFlingDistance.toFloat())
+
+        val indexDelta = (flingDistance / distancePerChild).roundToInt()
+
+        return currentItem.index + if (initialVelocity > 0) {
+            // If we're flinging forward, the current item is likely the same item as
+            // which the user started dragging
+            indexDelta
+        } else {
+            // If we're going backwards though, any backwards drag immediately goes back one item,
+            // so we need to cater for that by adding 1 to the result
+            (indexDelta + 1).coerceAtMost(lazyListState.layoutInfo.totalItemsCount - 1)
+        }
+    }
+
+    /**
+     * Computes an average pixel value to pass a single child.
+     *
+     * Returns a negative value if it cannot be calculated.
+     *
+     * @return A float value that is the average number of pixels needed to scroll by one view in
+     * the relevant direction.
+     */
+    private fun computeDistancePerChild(): Float {
+        if (lazyListState.layoutInfo.visibleItemsInfo.isEmpty()) return -1f
+
+        val minPosView = lazyListState.layoutInfo.visibleItemsInfo
+            .minByOrNull { it.offset } ?: return -1f
+        val maxPosView = lazyListState.layoutInfo.visibleItemsInfo
+            .maxByOrNull { it.offset + it.size } ?: return -1f
+
+        val start: Int = min(minPosView.offset, maxPosView.offset)
+        val end: Int = max(minPosView.offset + minPosView.size, maxPosView.offset + maxPosView.size)
+        val distance = end - start
+
+        return when {
+            distance != 0 -> distance.toFloat() / lazyListState.layoutInfo.visibleItemsInfo.size
+            else -> -1f
+        }
+    }
+
+    private fun determineTargetIndexForSpring(
+        initialVelocity: Float,
+        currentItemInfo: LazyListItemInfo,
+    ): Int {
+        return if (initialVelocity.absoluteValue > 1) {
+            // If the velocity isn't zero, spring in the relevant direction
+            if (initialVelocity > 0) currentItemInfo.index + 1 else currentItemInfo.index
+        } else {
+            // Otherwise we look at the current offset, and spring to whichever is closer
+            val snapOffset = snapOffsetForItem(lazyListState.layoutInfo, currentItemInfo)
+            if (currentItemInfo.offset < snapOffset - (currentItemInfo.size / 2)) {
+                currentItemInfo.index + 1
+            } else currentItemInfo.index
+        }.coerceIn(0, lazyListState.layoutInfo.totalItemsCount - 1)
+    }
+
+    private suspend fun ScrollScope.performSpringFling(
```

Comment:
this and performDecayFling share quite a lot of code, right? maybe we can try to extract some reusable parts?

Reply:
Yep, it took a bit of scope gymnastics but managed to get them sharing most of their impl